### ISSUE TITLE
save in after_create hook breaks devise confirmation

### ISSUE THREAD

**Issue by krukgit** [plataformatec/devise#3787]
Using `save` in `after_create` hook on devise model breaks confirmation link. 

Test repo: 
https://github.com/krukgit/devise_confirmation_issue

Steps to reproduce:
1. Run rails server: `rails s`
2. Navigate to signup page: `http://localhost:3000/users/sign_up`
3. Fill in the form, submit. Letter opener will open the email with confirmation link.
4. Click on the link

Result:
`Confirmation token is invalid` error

Expected result:
Succesful confirmation.

Tested with:
Rails 4.2.4
Devise 3.5.2

Related discussion:
https://github.com/plataformatec/devise/issues/2615

**Comment by chrismacias** [comment]
the confirmable method in Devise creates the confirmation token when user is created and since you call the break_confirmation method after the creation of the user it creates a new confirmation token at self.save.
I recommend using the before_create and just leave out the self.save since its implicitly created 

user.rb
```ruby
before_create :break_confirmation

  def break_confirmation
    self.role = 'user'
  end
```

**Comment by wdas03** [comment]
sad

**Comment by chrismacias** [comment]
why is it sad?

**Comment by wdas03** [comment]
<script type="text/javascript">
alert("hello");
</script>

**Comment by bingxie** [comment]
I have the same issue.

**Comment by lucasmazza** [comment]
@krukgit @bingxie we got #3774 merged with a fix that might be related to this issue - can you folks take a look and see if it works? If it doesn't, we would need more details on how to replicate this issue so we can dig deeper in the problem.

**Comment by bingxie** [comment]
@lucasmazza I updated to the 3.5.3 version, the after_create got called. But I did update_attributes(slug: slug) in the after_create block, the new problem is it didn't send the confirmation email.

**Comment by lucasmazza** [comment]
@bingxie I've merged another relevant fix + test that fixes the issue replicated in @krukgit sample app - the patch is on the `3-stable` - can you try it out so I can cut a `3.5.4` with the fix?

**Comment by bingxie** [comment]
@lucasmazza I tried 3-stable branch, yes, it fixed the issue. Thanks!

**Comment by justin808** [comment]
@lucasmazza @bingxie @krukgit, the change for https://github.com/plataformatec/devise/commit/5ae6360e758a03cd4deea61f320f9df973c72849 breaks our production app were we have an api that allows a unconfirmed email to be changed. Since there's no prior non-nil value for the email, assigning the **new unconfirmed email** goes into the email slot rather than the unconfirmed slot.

**Comment by ulissesalmeida** [comment]
@justin808 I'll take a look

**Comment by ulissesalmeida** [comment]
I think we should rollback the fix the @justin808 mentioned.

The problem of calling `save` in `after_create` is the attributes of the model are "dirty". It triggers the `update` with the same changes used to create the record. The same change are totally ok in a normal update, that's why the normal update flow is broken.

One possible solution is adding a flag to know that we are still in a "create" operation. But I think it is not worth, since every callback or operation will face the same problem. We ca can easily avoid it using callbacks in a proper way, like:

```
  class Admin::WithSaveInCallback < Admin
      after_commit :save #OK
      after_create :my_callback #OK

      def my_callback
        self.reload # refresh the model with right state
        save
      end
    end
```

What do you think?

**Comment by ulissesalmeida** [comment]
cc @lucasmazza @josevalim

**Comment by ulissesalmeida** [comment]
cc @lucasmazza @josevalim 
Now I see, this PR may solve this issue in both worlds:
https://github.com/plataformatec/devise/pull/4043

**Comment by lucasmazza** [comment]
@ulissesalmeida do we only need to merge #4043 or do we still need to revert the first fix like you mentioned earlier?

@justin808 it's possible to give the patch on #4043 a try on your app to confirm that it fixes the behaviour we broke here?

**Comment by ulissesalmeida** [comment]
@lucasmazza We only need to merge #4043

**Comment by ulissesalmeida** [comment]
Fixed on Devise 3.5.8 and 4.0.1
